### PR TITLE
Fix circular ref in dependency resolution

### PR DESCRIPTION
**Pull request by goetas**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | -

#### Summary

Ensure dependency resolution works even when using as dependency the default service definition

### CODE REVIEW

File: tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php]:
```suggestion
        $di->setDefinition(MigrationFactory::class, static function (DependencyFactory $innerDi) use ($migration): MigrationFactory {
```

File: tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php]: What does `di` mean here? Dependency injection? Shouldn't it be `$dependencyFactory`?

File: tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php]: Where is the reference in that test?

**Review by greg0ire — COMMENTED** [review]
Sorry, I don't really understand what this PR does. I see that we avoid calling a factory for something that we already are "resolving", but I don't understand what "the default service definition" is, or where the circular dependency is in the test.

**Comment by goetas** [comment]
@greg0ire thanks for the review. I think that I should improve the naming of the tests and some better description.

What is happening is the following.

You might want to provide a custom migration factory, but to avoid re implementing some of the things that doctrine already does well (providing the connection to the migration as example), what you wand to do is to **decorate** the current migration factory (otherwise you are forced to implement also the default migration factory rather than decorating it).


The code should look to something as follows:

```php

$dependencyFactory->setDefinition(MigrationFactory::class, function() use ($dependencyFactory) {
    $originalMigrationFactory = $dependencyFactory->getMigrationFactory();

    return new MyMigrationFactory($originalMigrationFactory);
 });
```

With the `3.0` version the provided code goes in a recursive loop. Mainly becase the ` $dependencyFactory->getMigrationFactory();` will try to invoke again the closure passed to `setDefinition`, going in recursion...

My change consists in avoiding to call the closure passed to `setDefinition` and to use the "default" implementation provided by the dependency factory.

In this way is possible to do proper service decoration.

**Comment by stof** [comment]
An alternative API that could be simpler would be to make the callable building the custom dependency receive a `$getOriginal` callable as argument that would return the original one.

File: tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php
Comment (stof) [inline tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php]: why do we need this exactly ? This creates a confusing API IMO

**Comment by goetas** [comment]
@stof 

> An alternative API that could be simpler would be to make the callable building the custom dependency receive a $getOriginal callable as argument that would return the original one.

how would that play with the symfony dependency injection container?

The proposed solution allows you to do something as:


```yaml
doctrine_migrations:
    services:
        Doctrine\Migrations\Version\MigrationFactory: 'doctrine.migrations.migrations_factory'
       
services:
  doctrine.migrations.migrations_factory:
      class: Doctrine\Migrations\Version\MigrationFactory
      factory: [@doctrine.migrations.dependency_factory'', 'getMigrationFactory']


  my_migrations_factory:
    class: Something
    decorates: doctrine.migrations.migrations_factory
   arguments:
      - '@my_migrations_factory.inner'

```

**Comment by goetas** [comment]
IMO, in the long term the whole dependency factory class could be replaced by some third party dependency injection container.... (the hard choice is to choose one as it should be compatible with all the other DICs).

I'm not aware on the status of https://github.com/Roave/psr-container-doctrine (cc @asgrim), but could be an option.

File: tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php
Comment (goetas) [reply tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php]: This is just because I was lazy to create some service that effectively decorates the one returned by `getMigrationFactory()`. Will create one and use the return value.

**Comment by asgrim** [comment]
@goetas `roave/psr-container-doctrine` is active (time permitting) :+1:

**Comment by goetas** [comment]
Thanks @asgrim for the fast feedback. I'm not familiar with the repo, but would it be able to "cooperate" with the symfony container and allows something similar to https://github.com/doctrine/migrations/pull/1054#issuecomment-694094516 (service decoration using services available in the symfony's container)?

**Comment by asgrim** [comment]
@goetas if you can point a service at an "invokable" factory class in Symfony's container, then yep, it should work; `psr-container-doctrine` is really just a set of factory classes

**Comment by stof** [comment]
@goetas AFAICT, your PR does not allow your snippet to work

**Comment by goetas** [comment]
@stof https://github.com/doctrine/DoctrineMigrationsBundle/pull/368 is using exactly the same strategy to provide container aware migrations by default in the bundle... and it seemed working...

**Comment by stof** [comment]
@goetas no. You are missing the extra code where it re-injects its own service in the dependency factory.

**Comment by goetas** [comment]
> @goetas no. You are missing the extra code where it re-injects its own service in the dependency factory.
This part of code does that:

```yaml
doctrine_migrations:
    services:
        Doctrine\Migrations\Version\MigrationFactory: 'doctrine.migrations.migrations_factory'
``` 

And is handled in https://github.com/doctrine/DoctrineMigrationsBundle/blob/880527161a1c88c8ccde558105441a9016f49650/DependencyInjection/DoctrineMigrationsExtension.php#L72

**Comment by massimilianobraglia** [comment]
Hey there, any news on this one?

**Comment by goetas** [comment]
I have updated the tests for this pull request with a (hopefully) more self explanatory test case.

File: lib/Doctrine/Migrations/DependencyFactory.php
Comment (greg0ire) [inline lib/Doctrine/Migrations/DependencyFactory.php]:
```suggestion
    /** @psalm-var array<string, bool> */
```

File: tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php]:
```suggestion
    /** @psalm-var class-string<CanItBeAnyClassOrShouldItBeASpecificType> */
```

File: tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php]: You will need to adapt my suggestion :laughing:

Comment (goetas) [reply tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php]: :exploding_head:

Comment (goetas) [reply tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php]: done

File: tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php
Comment (greg0ire) [inline tests/Doctrine/Migrations/Tests/Stub/CustomClassNameMigrationFactory.php:27]:
```suggestion
    /**
     * @psalm-param class-string<AbstractMigration> $migrationClassName
     */
    public function createVersion(string $migrationClassName): AbstractMigration
```

**Comment by goetas** [comment]
yay, green

**Comment by greg0ire** [comment]
@stof can you please give this another review :pray: ?

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Stof seems busy so I will approve, it looks good to me but I really would have liked another review from him :grimacing:

**Comment by goetas** [comment]
I'm merging this as it will unblock https://github.com/doctrine/DoctrineMigrationsBundle/pull/368 https://github.com/doctrine/DoctrineMigrationsBundle/pull/364 (and allow to close https://github.com/doctrine/DoctrineMigrationsBundle/issues/352)

**Pull request merged**
